src/request/change-password/ChangePassword.js

```diff
@@ -0,0 +1,211 @@
+            let colorClass = '';
+            if (this._key.secret instanceof Nimiq.Entropy) {
+                const color = Iqons.getBackgroundColorIndex(
+                    new Nimiq.Address(
+                        // use color of first address as loginFile color
+                        this._key.deriveAddress(Constants.DEFAULT_DERIVATION_PATH).serialize(),
+                    ).toUserFriendlyAddress(),
+                );
+                const colorString = LoginFile.CONFIG[color].name;
+                colorClass = `nq-${colorString}-bg`;
+            }
+            this._loginFileIcon.lock(colorClass);
+            this._progressIndicator.setStep(3);
+        });
+
+        this._passwordSetter.on(PassphraseSetterBox.Events.SUBMIT, /** @param {string} password */ async password => {
+            const passwordBytes = Utf8Tools.stringToUtf8ByteArray(password);
+
+            await KeyStore.instance.put(this._key, passwordBytes);
+
+            if (this._key.secret instanceof Nimiq.PrivateKey) {
+                this._resolve({ success: true });
+                return;
+            }
+
+            // Prepare Login File for download
+
+            const firstAddress = new Nimiq.Address(
+                this._key.deriveAddress(Constants.DEFAULT_DERIVATION_PATH).serialize(),
+            );
+
+            const encryptedEntropy = await this._key.secret.exportEncrypted(passwordBytes);
+
+            downloadLoginFile.setEncryptedEntropy(
+                /** @type {Nimiq.SerialBuffer} */ (encryptedEntropy),
+                firstAddress,
+            );
+
+            downloadLoginFile.on(DownloadLoginFile.Events.DOWNLOADED, () => {
+                this._resolve({ success: true });
+            });
+            window.location.hash = ChangePassword.Pages.DOWNLOAD_FILE;
+        });
+        this._passwordSetter.on(PassphraseSetterBox.Events.NOT_EQUAL, () => this._loginFileIcon.unlock());
+    }
+
+    async run() {
+        this._passwordGetter.reset();
+        window.location.hash = ChangePassword.Pages.ENTER_PASSWORD;
+        if (TopLevelApi.getDocumentWidth() > Constants.MIN_WIDTH_FOR_AUTOFOCUS) {
+            this._passwordGetter.focus();
+        }
+    }
+
+    /**
+     * @param {string} password
+     */
+    async _unlock(password) {
+        TopLevelApi.setLoading(true);
+        const passwordBytes = password ? Utf8Tools.stringToUtf8ByteArray(password) : undefined;
+        /** @type {Key?} */
+        let key = null;
+        try {
+            key = await KeyStore.instance.get(this._request.keyInfo.id, passwordBytes);
+        } catch (e) {
+            if (e.message === 'Invalid key') {
+                TopLevelApi.setLoading(false);
+                this._passwordGetter.onPassphraseIncorrect();
+                return;
+            }
+            this._reject(new Errors.CoreError(e));
+            return;
+        }
+        if (!key) {
+            this._reject(new Errors.KeyNotFoundError());
```

Comment:
_Technically_, key-existance is already checked during request parsing. I know that this check is also for Typescript, but it's still redundant and this code should never get triggered. I think we should make this error into something like a `ShouldNeverHappenError`, so that we are alerted when unexpected code paths are triggered. @nibhar?

Reply:
Maybe it got lost in the meantime, which `shouldNeverHappen` of course.